### PR TITLE
add aws:SecureTransport boolean condition to bucket policy ( #380 )

### PR DESCRIPTION
**Pull request by kuenishi**
#380

### CODE REVIEW

File: src/riak_cs_s3_policy.erl [hunk 1]
```diff
@@ -326,6 +328,16 @@ eval_date_eq(_, _) -> throw(not_supported).
 eval_date_lt(_, _) -> throw(not_supported).
 eval_date_lte(_, _) -> throw(not_supported).
 
+-spec eval_bool(#wm_reqdata{}, [{'aws:SecureTransport', boolean()}]) -> boolean().
+eval_bool(_Req, []) ->  undefined;
+
+eval_bool(#wm_reqdata{scheme=Scheme} = _Req, Conds) ->
+    {'aws:SecureTransport', Boolean} = hd(lists:reverse(Conds)),
```
Comment (shino) [inline src/riak_cs_s3_policy.erl]: hd + reverse is just lists:last/1 ?

File: src/riak_cs_s3_policy.erl [hunk 2]
```diff
@@ -551,7 +563,10 @@ condition_({<<"aws:CurrentTime">>, Bin}) when is_binary(Bin) ->
     {'aws:CurrentTime', Bin};
 condition_({<<"aws:EpochTime">>, Int}) when is_integer(Int) andalso Int >= 0 ->
     {'aws:EpochTime', Int};
-condition_({<<"aws:SecureTransport">>, Bool}) when is_atom(Bool) ->
+condition_({<<"aws:SecureTransport">>, MaybeBool}) ->
+    %% TODO: if this doesn't match return code like malformed condition -
+    %% while AWS never returns error in failing. Needs decision.
```
Comment (shino) [inline src/riak_cs_s3_policy.erl:584]: When set policy with Condition like this:

```
[snip]
      "Condition": {
        "Bool": {
          "aws:SecureTransport": t
        }
      },
[snip]
```

S3 responds with 400 MalformedPolicy.

```
$ s3cmd -v setpolicy policy_https_only_put.txt s3://shino.shun.test-https-only
ERROR: S3 error: 400 (MalformedPolicy): JSON parsing error
```

File: src/riak_cs_s3_policy.erl [hunk 3]
```diff
@@ -706,4 +754,45 @@ parse_arn_test()->
...
+      "   }"
+      "  }"
+      " ]"
+      "}" >>.
+
+
+secure_transport_test()->
+    JsonPolicy0 = sample_statement_1(),
+    Policy = policy_from_json(JsonPolicy0),
+    Req = #wm_reqdata{peer="192.168.0.1", scheme="https"},
```
Comment (shino) [inline src/riak_cs_s3_policy.erl]: eunit fails:

```
$ ./rebar eunit suites=riak_cs_s3_policy skip_deps=true
==> rel (eunit)
==> riak_cs (eunit)
riak_cs_s3_policy: secure_transport_test...*failed*
::{case_clause,"https"}


=======================================================
  Failed: 1.  Skipped: 0.  Passed: 17.
Cover analysis: /home/shino/work/basho/git/riak_cs/.eunit/index.html
ERROR: One or more eunit tests failed.
ERROR: eunit failed while processing /home/shino/work/basho/git/riak_cs: rebar_abort
```

`"https"` should be atom?

File: src/riak_cs_s3_policy.erl [hunk 4]
```diff
@@ -706,4 +754,45 @@ parse_arn_test()->
...
+    Access = #access_v1{method='GET', target=object, id="spam/ham/egg",
+                        req = Req, bucket= <<"test">>},
+    ?assert(eval(Access, Policy)),
+    % io:format(standard_error, "~w~n", [Policy]),
+    Access2 = Access#access_v1{req=Req#wm_reqdata{scheme="http"}},
```
Comment (shino) [inline src/riak_cs_s3_policy.erl]: ditto

File: src/riak_cs_s3_policy.erl (hunk 1 above)
Comment (kuenishi) [inline src/riak_cs_s3_policy.erl]: right. will be fixed.

File: src/riak_cs_s3_policy.erl (hunk 2 above)
Comment (kuenishi) [inline src/riak_cs_s3_policy.erl:584]: I think it works:
https://github.com/basho/riak_cs/blob/master/src/riak_cs_wm_bucket_policy.erl#L102

File: src/riak_cs_s3_policy.erl (hunk 3 above)
Comment (kuenishi) [inline src/riak_cs_s3_policy.erl]: yep, will fix

File: src/riak_cs_s3_policy.erl (hunk 4 above)
Comment (kuenishi) [inline src/riak_cs_s3_policy.erl]: DITTO

File: src/riak_cs_s3_policy.erl (hunk 2 above)
Comment (kuenishi) [inline src/riak_cs_s3_policy.erl:584]: no, it didn't work - due to mochijson2 throwing unmached case. will be fixed.

**Comment by kuenishi** [comment]
OK, pushed fixes - test and review please!

File: src/riak_cs_s3_policy.erl (hunk 1 above)
Comment (shino) [inline src/riak_cs_s3_policy.erl]: Not yet?

File: src/riak_cs_s3_policy.erl [hunk 5]
```diff
@@ -145,12 +148,18 @@ reqdata_to_access(RD, Target, ID) ->
 
 -spec policy_from_json(JSON::binary()) -> policy().
```
Comment (shino) [inline src/riak_cs_s3_policy.erl]: return type is changed to `{ok, policy()} | {error, term()}`

File: src/riak_cs_s3_response.erl [hunk 6]
```diff
@@ -43,6 +43,7 @@ error_message(admin_key_undefined) -> "Please reduce your request rate.";
 error_message(admin_secret_undefined) -> "Please reduce your request rate.";
 error_message(bucket_owner_unavailable) -> "The user record for the bucket owner was unavailable. Try again later.";
 error_message(econnrefused) -> "Please reduce your request rate.";
+error_message(malformed_policy) -> "JSON parsing error";
```
Comment (shino) [inline src/riak_cs_s3_response.erl]: This is not bad, but does the atom with suffix, such as`malformed_policy_json` , suit better?
The reason is some atoms below here have suffixes.

File: src/riak_cs_s3_policy.erl [hunk 7]
```diff
@@ -145,12 +148,18 @@ reqdata_to_access(RD, Target, ID) ->
...
-    Stmts = lists:map(fun({struct,S})->statement_from_pairs(S, #statement{}) end, Stmts0),
-    ?POLICY{id=ID, version=Version, statement=Stmts}.
+    case catch(mochijson2:decode(JSON)) of
+        {struct, Pairs} ->
+            Version = proplists:get_value(<<"Version">>, Pairs),
+            ID      = proplists:get_value(<<"Id">>, Pairs),
+            Stmts0  = proplists:get_value(<<"Statement">>, Pairs),
+            Stmts = lists:map(fun({struct,S})->statement_from_pairs(S, #statement{}) end, Stmts0),
+            {ok, ?POLICY{id=ID, version=Version, statement=Stmts}};
+        {error, _Reason} = E->
```
Comment (shino) [inline src/riak_cs_s3_policy.erl]: One space before `->`

File: src/riak_cs_s3_policy.erl [hunk 8]
```diff
@@ -706,4 +763,45 @@ parse_arn_test()->
     {ok, ARNS2} = parse_arns(List2),
     ?assertEqual(List2, print_arns(ARNS2)).
 
+sample_statement_1()->
```
Comment (shino) [inline src/riak_cs_s3_policy.erl]: I like some descriptive name, e.g. deny_plain_http_statement.

**Comment by shino** [comment]
Response code diffrence is filed as a separate issue #403

**Comment by shino** [comment]
+1 to merge!

**Pull request merged**